03-spatial-joins.ipynb

```diff
@@ -239,6 +239,92 @@
     "joined['continent'].value_counts()"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "\n",
+    "**Exercise**: Which rivers pass within 1 degree of Paris? And for Hong Kong?\n",
+    "\n",
+    "Tip: `rivers` contains empty geometries, make sure to get rid of them before anything else\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<!--\n",
+    "paris = cities.query('name == \"Paris\"')\\\n",
+    "              ['geometry']\\\n",
+    "              .squeeze()\\\n",
+    "              .buffer(1)\n",
+    "\n",
+    "river_lines = rivers[rivers.geometry.notna()]\n",
+    "\n",
+    "river_lines[river_lines.crosses(paris)]\n",
+    "\n",
+    "hk = cities.query('name == \"Hong Kong\"')\\\n",
+    "              ['geometry']\\\n",
+    "              .squeeze()\\\n",
+    "              .buffer(1)\n",
+    "\n",
+    "river_lines = rivers[rivers.geometry.notna()]\n",
+    "\n",
+    "river_lines[river_lines.crosses(hk)]\n",
+    "-->"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "\n",
+    "**Exercise** [PRO]: Find all European cities within 1 degre of at least one river\n",
+    "\n",
+    "\n",
+    "This requires:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "1. Obtaining a list of European cities\n",
+    "1. Creating a `GeoDataFrame` with all river lines\n",
+    "1. For every European city, checking whether at least one river \"crosses\" the area within 1 degree\n",
+    "1. If so, saving the record of a city"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<!--\n",
+    "eu_cities = joined.query('continent == \"Europe\"')\n",
+    "river_lines = rivers[rivers.geometry.notna()]\n",
+    "\n",
+    "answer = []\n",
+    "for id, city in eu_cities.iterrows():\n",
+    "    buf = city['geometry'].buffer(1)\n",
+    "    rivers_crossing = river_lines.crosses(buf).sum()\n",
+    "    if rivers_crossing > 0:\n",
+    "        answer.append(city)\n",
+    "answer = geopandas.GeoDataFrame(answer, \n",
```

Comment:
I would personally do it with a function and `apply`. It is maybe more complex in the end, but a good pattern to show IMO. 
Or first with loop, and then as exercise to convert it to apply